File: qa/gdk/normalize-executable-shebangs.py
"""Normalize committed CRLF scripts without touching binary or linked inputs."""
import os
from pathlib import Path
import sys


def normalize_file(path):
    if path.is_symlink():
        return 0
    with path.open("rb") as source:
        prefix = source.read(2)
        if prefix != b"#!":
            return 0
        original = prefix + source.read()
    converted = original.replace(b"\r\n", b"\n")
    if converted == original:
        return 0
    path.write_bytes(converted)
    return 1


def normalize(roots):
    changed = 0
    for root in map(Path, roots):
        if root.is_symlink() or not root.exists():
            continue
        if root.is_file():
            changed += normalize_file(root)
            continue
        for directory, children, files in os.walk(root, followlinks=False):
            children[:] = [name for name in children
                           if name != "node_modules" and not (Path(directory) / name).is_symlink()]
            for filename in files:
                changed += normalize_file(Path(directory) / filename)
    return changed


if __name__ == "__main__":
    print(f"Normalized CRLF shebang files: {normalize(sys.argv[1:])}")

File: qa/gdk/normalize-executable-shebangs.rb
# frozen_string_literal: true

require 'find'

normalized = 0

normalize_file = lambda do |path|
  bytes = File.binread(path)
  next unless bytes.start_with?('#!')

  converted = bytes.gsub("\r\n", "\n")
  next if converted == bytes

  File.binwrite(path, converted)
  normalized += 1
end

ARGV.each do |root|
  next unless File.exist?(root)

  if File.file?(root)
    normalize_file.call(root) unless File.symlink?(root)
    next
  end

  Find.find(root) do |path|
    stat = File.lstat(path)
    if stat.symlink?
      Find.prune
      next
    end
    if stat.directory?
      Find.prune if File.basename(path) == 'node_modules'
      next
    end
    normalize_file.call(path) if stat.file?
  end
end

puts "Normalized CRLF shebang files: #{normalized}"
